sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,311 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_plic.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+// Interrupt Pending register offsets from the PLIC base
+#define RV_PLIC_IP0 0x0u
+#define RV_PLIC_IP1 0x4u
+
+// Level/Edge register offsets from the PLIC base
+#define RV_PLIC_LE0 0x8u
+#define RV_PLIC_LE1 0xCu
+
+// Priority register offsets from the PLIC base
+#define RV_PLIC_PRIO0 0x10u
+
+// Interrupt Enable register offsets from the PLIC base
+#define RV_PLIC_IE00 0x200u
+#define RV_PLIC_IE01 0x204u
+
+// Threshold register offsets from the PLIC base
+#define RV_PLIC_THRESHOLD0 0x208u
+
+// Interrupt Claim register offsets from the PLIC base
+#define RV_PLIC_CC0 0x20cu
+
+// The highest interrupt priority
+#define RV_PLIC_MAX_PRIORITY 0x3u
+
+// Peripheral interrupt ID ranges
+#define PLIC_FIRST_IRQ_ID kDifPlicGpio0
+#define PLIC_LAST_IRQ_ID kDifPlicNmiGenEsc3
+#define PLIC_GPIO_FIRST_IRQ_ID kDifPlicGpio0
+#define PLIC_GPIO_LAST_IRQ_ID kDifPlicGpio31
+#define PLIC_UART_FIRST_IRQ_ID kDifPlicUartTxWatermark
+#define PLIC_UART_LAST_IRQ_ID kDifPlicUartRxParityErr
+#define PLIC_SPI_DEVICE_FIRST_IRQ_ID kDifPlicSpiDeviceRxF
+#define PLIC_SPI_DEVICE_LAST_IRQ_ID kDifPlicSpiDeviceTxUnderflow
+#define PLIC_FLASH_CONTROL_FIRST_IRQ_ID kDifPlicFlashCtrlProgEmpty
+#define PLIC_FLASH_CONTROL_LAST_IRQ_ID kDifPlicFlashCtrlOpError
+#define PLIC_HMAC_FIRST_IRQ_ID kDifPlicHmacDone
+#define PLIC_HMAC_LAST_IRQ_ID kDifPlicHmacErr
+#define PLIC_ALERT_HANDLER_FIRST_IRQ kDifPlicAlertHandlerClassA
+#define PLIC_ALERT_HANDLER_LAST_IRQ kDifPlicAlertHandlerClassD
+#define PLIC_NMI_GEN_FIRST_IRQ kDifPlicNmiGenEsc0
+#define PLIC_NMI_GEN_LAST_IRQ kDifPlicNmiGenEsc3
+
+// Peripheral IRQ sources that interrupt the Ibex core
+#define PLIC_IBEX_SOURCE_FIRST_IRQ kDifPlicGpio0
+#define PLIC_IBEX_SOURCE_LAST_IRQ kDifPlicNmiGenEsc3
```

Comment:
I'm not sure that these FIRST/LAST macros do anything to improve readability, especially seeing as many are only used once...

Reply:
I agree with @gkelly on this one. I think this is just a bit more descriptive, I agree that it would be more useful if there were multiple places where these are used.

There is another related comment further down, and in case we decide to keep the current range checking - I would like to keep these defines, but if we go for a switch style range checking that @mcy suggested, then I will remove these.

Reply:
Went with @alphan suggestion to use an array of structs